src/plonk/lookup/prover.rs

```diff
@@ -0,0 +1,621 @@
+}
+
+impl<'a, C: CurveAffine> Permuted<'a, C> {
+    /// Given a Lookup with input columns, table columns, and the permuted
+    /// input column and permuted table column, this method constructs the
+    /// grand product polynomial over the lookup. The grand product polynomial
+    /// is used to populate the Product<C> struct. The Product<C> struct is
+    /// added to the Lookup and finally returned by the method.
+    pub(in crate::plonk) fn commit_product<HBase: Hasher<C::Base>, HScalar: Hasher<C::Scalar>>(
+        self,
+        pk: &ProvingKey<C>,
+        params: &Params<C>,
+        theta: ChallengeTheta<C::Scalar>,
+        beta: ChallengeBeta<C::Scalar>,
+        gamma: ChallengeGamma<C::Scalar>,
+        transcript: &mut Transcript<C, HBase, HScalar>,
+    ) -> Result<Committed<'a, C>, Error> {
+        // Goal is to compute the products of fractions
+        //
+        // (\theta^{m-1} a_1(\omega^i) + \theta^{m-2} a_2(\omega^i) + ... + a_m(\omega^i) + \beta)(\theta^{m-1} s_1(\omega^i) + \theta^{m-2} s_2(\omega^i) + ... + s_m(\omega^i) + \gamma)/
+        // (a'(\omega^i) + \beta) (s'(\omega^i) + \gamma)
+        //
+        // where a_j(X) is the jth input column in this lookup,
+        // where a'(X) is the compression of the permuted input columns,
+        // s_j(X) is the jth table column in this lookup,
+        // s'(X) is the compression of the permuted table columns,
+        // and i is the ith row of the column.
+        let mut lookup_product = vec![C::Scalar::zero(); params.n as usize];
+        // Denominator uses the permuted input column and permuted table column
+        parallelize(&mut lookup_product, |lookup_product, start| {
+            for ((lookup_product, permuted_input_value), permuted_table_value) in lookup_product
+                .iter_mut()
+                .zip(self.permuted_input_column[start..].iter())
+                .zip(self.permuted_table_column[start..].iter())
+            {
+                *lookup_product = (*beta + permuted_input_value) * &(*gamma + permuted_table_value);
+            }
+        });
+
+        // Batch invert to obtain the denominators for the lookup product
+        // polynomials
+        lookup_product.iter_mut().batch_invert();
+
+        // Finish the computation of the entire fraction by computing the numerators
+        // (\theta^{m-1} a_1(\omega^i) + \theta^{m-2} a_2(\omega^i) + ... + \theta a_{m-1}(\omega^i) + a_m(\omega^i) + \beta)
+        // * (\theta^{m-1} s_1(\omega^i) + \theta^{m-2} s_2(\omega^i) + ... + \theta s_{m-1}(\omega^i) + s_m(\omega^i) + \gamma)
+        parallelize(&mut lookup_product, |product, start| {
+            for (i, product) in product.iter_mut().enumerate() {
+                let i = i + start;
+
+                // Compress unpermuted input columns
+                let mut input_term = C::Scalar::zero();
+                for unpermuted_input_column in self.unpermuted_input_columns.iter() {
+                    input_term *= &theta;
```

Comment:
Hmm, clippy also seems to be complaining in a similar way about a bunch of code not added in this PR. Must be due to some other recent refactoring.

Reply:
yeah, it seems like clippy isn't using our `Deref` impl for `ChallengeScalar`.

Reply:
There are many issues with clippy that we'll be resolving once we've emptied the PR queue, because we introduced clippy lints in CI during a lot of ongoing development.